### ISSUE TITLE
Restrict operator privileges

### ISSUE THREAD

**Issue by tgeoghegan** [abetterinternet/prio-server#1170]
Our opsec could be improved considerably. Specifically, we should be deploying to production from some secure and ideally automated environment like a Github Action, and operators should have to opt-into dangerous cloud platform credentials by assuming an AWS IAM role or a GCP service account (i.e., our human users should have readonly privileges at best).

**Comment by tgeoghegan** [comment]
I have been tinkering with this in EKS/AWS and have concocted a Scheme.

This ticket is about a few different things. What I've worked out this afternoon is a way to have operators use read-only credentials for AWS accounts and EKS clusters by default, and use separate credentials and authentication contexts to assume privileges that allow modification or destruction of resources. This scheme should also address #1031. It won't address #1165, but it does introduce an IAM indirection that should make it easier to do.

What this scheme does _not_ do is eliminate insider threats. The aim here is to minimize the risk of _accidental_ damage to deployed resources by requiring one extra hurdle to be leaped before dangerous credentials are available in an operator's shell. This is also AWS and EKS specific; something analogous will have to be worked out for GCP/GKE.

## AWS credentials

Currently, our operators have an IAM user in each AWS account with full admin privileges, via membership in an IAM group that is granted the FullAdmin policy. I propose to:

- remove all permissions and group memberships from IAM users (see note about the magic `system:masters` member below)
- create role `read-only-access` with policy `arn:aws:iam::aws:policy/ReadOnlyAccess`
- create role `admin-access` with policy `arn:aws:iam::aws:policy/AdministratorAccess`, and a role assumption policy requiring MFA
- attach policies to each IAM user allowing assumption of roles `read-only-access` and `admin-access`, the latter requiring MFA

Then, we have each operator configure their `~/.aws/config` with entries like:

```
[profile my-iam-user]
region = us-west-2

[profile my-iam-user-ro]
role_arn = arn:aws:iam::<ACCOUNT#>:role/read-only-access
source_profile = my-iam-user
role_session_name = <useful string for audit logs>

[profile my-iam-user-admin]
role_arn = arn:aws:iam::<ACCOUNT#>:role/admin-access
source_profile = my-iam-user
role_session_name = <useful string for audit logs>
mfa_serial = arn:aws:iam::<ACCOUNT#>:mfa/<mfa-device-name>
```

With this in place, operators can do `aws --profile=my-iam-user-ro <command>` to safely view resources. If they use `aws --profile=my-iam-user-admin <command>`, they get prompted by the AWS CLI to enter an MFA token code, which should prevent a lot of accidental destructive actions.

## EKS cluster credentials

EKS cluster credentials are related to AWS IAM but separate. The relationship between the two is encoded in the [`kube-system/aws-auth` configmap](https://docs.aws.amazon.com/eks/latest/userguide/add-user-role.html). We will use `aws-auth` to grant the `admin-access` IAM role permission to use the `system:masters` group (defined in the `kube-system:cluster-admins` `ClusterRoleBinding), and grant the `read-only-access` IAM role permission to use the `read-only` group. We define the `read-only` group with a `ClusterRoleBinding` referencing the built-in `view` `ClusterRole`:

```
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: read-only
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: view
subjects:
- apiGroup: rbac.authorization.k8s.io
  kind: Group
  name: read-only
```

With this in place, operators can use `aws eks update-kubeconfig` (along with a little bit of editing of `~/.kube/config` to create `kubectl` contexts that use the `read-only-access` IAM role or the `admin-access` IAM role to get to the appropriate Kubernetes group.

One gotcha here is that EKS has a surprising behavior where [the IAM user that creates an EKS cluster is invisibly and irrevocably added to `system:masters`](https://github.com/aws/containers-roadmap/issues/922). The only workaround would be to delete the IAM user that created the EKS account, once a new user suitably anointed with the ability to assume admin privileges is created.

This resolves #1031 because we now grant `system:masters` membership to an IAM role instead of to a list of IAM users. We can then manage the ability to assume the role outside of Terraform. Further, we can later implement #1165 more easily because each AWS account will now have roles defined in them which can be used with the switch roles feature.

**Comment by tgeoghegan** [comment]
This is quite a bit easier on the GCP side. GCP has canned roles `project/viewer` and `project/owner` that already integrate pretty well with Kubernetes RBAC. A user or service account with `project/owner` basically can edit anything in a GKE cluster in that project, and one with `project/viewer` gets reasonable readonly privileges on the clusters. So in particularly sensitive GCP projects like prod or staging, we should:

- remove `project/owner` from individual operator GCP users, replace with `project/viewers`
- create a new service account called `project-admin@<project-name>.iam.gserviceaccount.com`
- grant the ENPA ops Google Group `roles/iam.serviceAccountTokenCreator` on the `project-admin` SA

Then, operators will want to construct appropriate entries in their `~/.kube/config` for `gke-cluster/admin` and `gke-cluster/ro`. The former uses `gcloud --impersonate-service-account` (or `gcloud --configuration` with a config that sets up SA impersonation) and grants them admin/edit/delete permissions. The latter will use `gcloud` credentials for their user, and grants them view permissions.

**Comment by tgeoghegan** [comment]
Implementing the above changes would also require some updates to [our onboarding docs](https://github.com/abetterinternet/docs/blob/main/prio-server/onboarding.md) to provide guidance on setup of new AWS accounts or GCP projects, as well as how to configure `gcloud`, `aws` and `kubectl`.

**Comment by tgeoghegan** [comment]
One frustrating wrinkle here is that the `terraform` `aws` provider [doesn't interact well with MFA](https://github.com/hashicorp/terraform-provider-aws/issues/2420), in that it won't prompt the user for an MFA code during a `terraform` operation. The mitigation there would be to use `aws sts assume-role` to construct an `aws` profile with temporary credentials that can then be used normally by the `aws` provider. That can be documented and we can circulate a simple script that does it.

The other wrinkle -- or rather, lack of a wrinkle -- is that I can't see how to force the presentation of a second factor during service account impersonation on GCP. You can configure an SA to allow impersonation by some user or group of users, and you can require MFA to authenticate as a user, but you can't seem to require MFA again when impersonating a user. I was hoping to introduce a manual step there to make operators go "wait, did I really mean to impersonate the `project-admin` SA?" But at least now operators would have to explicitly `kubectl config use-context` into dangerous credentials, or use `gcloud --configuration=something-dangeous`.

**Comment by tgeoghegan** [comment]
Brandon and I chatted about this today and agreed to move forward with this. To wrap this up, I need to:

- [ ] document how to set up configs, contexts, credentials for `aws`, `gcloud` and `kubectl` in [the onboarding guide](https://github.com/abetterinternet/docs/tree/main/prio-server)
- [ ] document how to bring up a new Prio env in either EKS or GKE, describing account or project creation (respectively) along with the necessary roles, policies, service accounts etc. Possibly also provide a Terraform module that, given a new GCP project or AWS account, sets things up appropriately.
- [ ] make changes in `prio-server/terraform` Makefile to have it gracefully handle MFA-protected roles to work around https://github.com/hashicorp/terraform-provider-aws/issues/2420. In particular this needs to be conditional, because we won't bother requiring the MFA dance for devenvs, only prod and maybe staging.
- [ ] write up an implementation plan describing changes that will need to be made manually to existing prod and staging environments.

**Comment by tgeoghegan** [comment]
I've hit a snag here. AWS' switch roles feature makes it possible to log into the AWS web console while assuming some role, so you can login as the readonly role or the admin role, depending on what you're doing. GCP has no equivalent feature, so if we institute the service account impersonation scheme I described, you can't make modifications via the GCP console. I can think of two ways around this, neither of which I like much.

Instead of provisioning a `project-admin` service account, we could create distinct users in our Google Workspace for each operator (i.e., `janedoe@` and `janedoe-gcp-admin@`). One of them gets `project/viewer` and one of them gets `project/admin`. GCP's web console makes it pretty easy to switch between logged-in Google identities, at least. 

We could instead stick with the `project-admin` SA scheme and document that if you need to make modifications via the console, you first need to use `gcloud` to assume the `project-admin` SA and grant yourself `project/owner`.

**Comment by tgeoghegan** [comment]
Another snag: while messing with our Terraform to run it under a script that uses `aws sts assume-role` to get MFA'ed credentials, I noticed that while Brandon has moved all the key creation and manifest posting to `key-rotator`, we still need `deploy-tool` to set up a batch signing key and a manifest for the fake ingestion server. Since that fake ingestor is only used in test configs, I think we can tolerate leaking its batch signing key into Terraform state. I'm going to see about moving the generation of the fake ingestion server's resources purely into Terraform. This will allow me to scrub the notion of the `aws_profile` from Terraform, which gets us a hair closer to achieving this goal.

**Comment by tgeoghegan** [comment]
To smooth out the provisioning of IAM roles in AWS accounts, I decided to make a couple new Terraform projects, introduced by #1243.

`terraform/aws-bootstrap/management_account` creates a new AWS account, subordinate to the specified management account, and creates IAM policies enabling groups in the management account to assume roles in the member account via Switch Roles.

`terraform/aws-bootstrap/member_account` ensures that the roles exist in the member account and also creates ECR repositories for `prio-facilitator`, `prio-workflow-manager` and `prio-key-rotator`. The reason I did this in two separate TF projects is because I didn't want to be configuring an `aws` provider for the member account in the same module where that account is created, since Hashicorp strongly advises against this. Further, I expect to be moving the creation of the EKS cluster out of the existing TF project (for the same reason; using a `kubernetes` provider in the same project that creates the cluster is here-be-demons territory), so this project will be a natural place to move the existing `eks` module.

So with this, bringing up a new EKS prio-server cluster goes something like:

- `terraform apply` with `management_account`
- `terraform apply` with `member_account`
- `ENV=my-new-env make apply-bootstrap`
- `ENV=my-new-env make apply`

I anticipate doing something similar for GCP projects, but I am not yet sure whether it will require two TF modules as AWS did.
